File: dynamic_programming/delete_and_earn.cpp
class Solution {
public:
    // TC: O( mac( sizeof( nums ), sizeof( count ) ) )
    // SC: O( sizeof( count ) )
    int deleteAndEarn(vector<int>& nums)
    {
        std::vector<int> count( 10001 );
        for( int n : nums )
            ++count[ n ];
        
        int prev{ 0 };
        int use{ 0 };
        int avoid{ 0 };
        for( int n = 0; n < count.size(); ++n ) {
            if( !count[ n ] )
                continue;
            
            int val = std::max( use, avoid );
            if( prev == n - 1 ) {
                use = n * count[ n ] + avoid;
                avoid = val;
            } else {
                use = n * count[ n ] + val;
                avoid = val;
            }

            prev = n;
        }
        
        return std::max( use, avoid );
    }
};

File: dynamic_programming/delete_and_earn.rs
impl Solution {
    // TC: O( max( sizeof( nums ), sizeof( count ) )
    // SC: O( sizeof( count ) )
    pub fn delete_and_earn(nums: Vec<i32>) -> i32 {
        let mut count = vec![ 0; 10001 ];
        for n in 0..nums.len() {
            let index: usize = nums[ n ] as usize;
            count[ index ] = count[ index ] + 1;
        }
        
        let mut prev = 0;
        let mut used = 0;
        let mut avoid = 0;
        for n in 0..count.len() {
            if count[ n ] == 0 {
                continue;
            }
            
            let val = std::cmp::max( avoid, used );
            if prev == n - 1 {
                used = n * count[ n ] + avoid;
                avoid = val;
            } else {
                used = n * count[ n ] + val;
                avoid = val;
            }

            prev = n;
        }
        
        let res = std::cmp::max( avoid, used );
        return res as i32;
    }
}
